### PR TITLE
Test Flat + Curve

### PR DESCRIPTION
**Pull request by sentilesdal**
- rename files
- add hyperdrive test files
- add comments
- remove unnecessary paranthesis
- fix typo
- fix bug
- fix bug
- changes to test files
- add file to generate expected results

### CODE REVIEW

File: src/elfpy/pricing_models/base.py
Comment (jalextowle) [inline src/elfpy/pricing_models/base.py:208]: Please add a TODO here. In general $s \neq y + cz$ and we'll want to update this soon to use `s = market_state.lp_reserves`

File: src/elfpy/pricing_models/hyperdrive.py
Comment (jalextowle) [inline src/elfpy/pricing_models/hyperdrive.py:129]: We don't want to add the flat component to the market bonds as this introduces a convexity problem. The previous implementation was correct.

File: src/elfpy/pricing_models/yieldspace.py
Comment (jalextowle) [inline src/elfpy/pricing_models/yieldspace.py:356]: You might as well use the unicode symbol for tau since we already do this for mu.

File: src/elfpy/pricing_models/yieldspace.py
Comment (jalextowle) [inline src/elfpy/pricing_models/yieldspace.py:361]: Out is denominated in base not shares (just like how our smart contract will accept base rather than shares). The previous implementation was correct.

File: src/elfpy/pricing_models/hyperdrive.py
Comment (sentilesdal) [reply src/elfpy/pricing_models/hyperdrive.py:116]: Ah right, this is what we were talking about last week.  Thanks for the clarification.

File: src/elfpy/pricing_models/hyperdrive.py
Comment (sentilesdal) [reply src/elfpy/pricing_models/hyperdrive.py:129]: Ok thanks

File: src/elfpy/pricing_models/yieldspace.py
Comment (sentilesdal) [reply src/elfpy/pricing_models/yieldspace.py:361]: Ok look at this one again.  The next line of code, line 371 has:
```
              without_fee_or_slippage = (1 / spot_price) * market_state.share_price * d_shares
 ```
 
 since the output is base, we multiply by c since `x = c * z`

Comment (jalextowle) [reply src/elfpy/pricing_models/yieldspace.py:361]: Let's chat about this after all-hands.

File: src/elfpy/pricing_models/hyperdrive.py
Comment (sentilesdal) [reply src/elfpy/pricing_models/hyperdrive.py:116]: undid this

File: src/elfpy/pricing_models/base.py
Comment (sentilesdal) [reply src/elfpy/pricing_models/base.py:208]: added TODO

File: src/elfpy/pricing_models/hyperdrive.py
Comment (sentilesdal) [reply src/elfpy/pricing_models/hyperdrive.py:116]: also added a TODO to change `days=365` to the length of the term during setup.

File: src/elfpy/pricing_models/hyperdrive.py
Comment (sentilesdal) [reply src/elfpy/pricing_models/hyperdrive.py:129]: undid this.

File: src/elfpy/pricing_models/yieldspace.py
Comment (sentilesdal) [reply src/elfpy/pricing_models/yieldspace.py:356]: I made this change, but it was cumbersome and I can see grepping etc being cumbersome in the future.  If you have no strong feelings I could to a search and replace all to get rid of greek symbols with alphabet spellings instead.

Comment (sentilesdal) [reply src/elfpy/pricing_models/yieldspace.py:356]: in a follow up PR i mean

Comment (dpaiton) [reply src/elfpy/pricing_models/yieldspace.py:356]: agreed, I'd rather us spell it out (`mu`, `tau`, etc)

Comment (jalextowle) [reply src/elfpy/pricing_models/yieldspace.py]: I prefer $\mu$ and $\tau$, but I don't care enough to argue about it.

File: src/elfpy/pricing_models/hyperdrive.py
Comment (jalextowle) [inline src/elfpy/pricing_models/hyperdrive.py:121]: IMO, this change should be part of this effort. We should have tests with non-365 day terms that capture the problems caused by our current normalization (again, this won't manifest on this line).

File: src/elfpy/pricing_models/base.py
Comment (jalextowle) [inline src/elfpy/pricing_models/base.py:330]: Let's use `mu` or $\mu$ instead of `u`.

File: src/elfpy/pricing_models/yieldspace.py
Comment (jalextowle) [inline src/elfpy/pricing_models/yieldspace.py:360]: Let's pick $\tau$ or `tau` in this PR. I don't really care, but would be good to set the record straight. Let's do the same for picking between $\mu$ and `mu`.

File: tests/pricing_models/hyperdrive/hyperdrive_generate_expected_results.py
Comment (jalextowle) [inline tests/pricing_models/hyperdrive/hyperdrive_generate_expected_results.py:1]: I'm against merging unlinted files into the repo. It's fine to keep around a useful tool, but it needs to meet the repo's standards.

Comment (jalextowle) [reply tests/pricing_models/hyperdrive/hyperdrive_generate_expected_results.py:1]: This file needs to be cleaned up before I'll approve.

I also just want to make sure that this is not replacing having your eyes on the inputs and outputs to make sure they are reasonable. I used a calculator (`python -i`) when I was writing the original pricing model tests, so it's totally fine to use another tool; however, we just want to make sure that we're carefully considering what we're putting into our test cases as we go along. I'm not suggesting that you aren't doing this, but I just think it's worth calling out.

File: tests/pricing_models/yieldspace/test_calc_in_given_out.py
Comment (jalextowle) [inline tests/pricing_models/yieldspace/test_calc_in_given_out.py:172]: Let's make this change everywhere in this file and `test_calc_out_given_in`.

File: tests/pricing_models/hyperdrive/hd_test_calc_in_given_out.py
Comment (jalextowle) [inline tests/pricing_models/hyperdrive/hd_test_calc_in_given_out.py:57]: We're duplicating a lot of code between these test files. My preference is that we use common inputs and provide the different outputs as a dictionary keyed by the model's name. That way we keep everything in one file and all of the pricing models benefit when we add cases.

File: tests/pricing_models/hyperdrive/hd_test_calc_in_given_out.py
Comment (jalextowle) [inline tests/pricing_models/hyperdrive/hd_test_calc_in_given_out.py:588]: This will need to be uncommented before it's merged.

**Review by jalextowle — CHANGES_REQUESTED** [review]
I'll need to do a final review after this.

File: src/elfpy/pricing_models/hyperdrive.py
Comment (dpaiton) [reply src/elfpy/pricing_models/hyperdrive.py:121]: Agreed, although IMO the test PR should not include new features so I'd say make that its own PR

File: tests/pricing_models/hyperdrive/hyperdrive_generate_expected_results.py
Comment (sentilesdal) [reply tests/pricing_models/hyperdrive/hyperdrive_generate_expected_results.py:1]: Yeah totally.  I'm not actually running anything through a calculator, but I'm visually inspecting things to make sure they look approximately sane.  I think a lot of the value for these tests come from the fact that the output is generated by a function that has all the math inlined, so we can inspect that the implementation code is doing what we want it to do.

**Review by dpaiton — APPROVED** [review]
Looks good after you address Alex's remaining concerns (add the failure tests, fix `mu` etc, & move the test helper file out of the repo).

File: src/elfpy/pricing_models/base.py
Comment (sentilesdal) [reply src/elfpy/pricing_models/base.py:330]: I went with `mu`

File: src/elfpy/pricing_models/yieldspace.py
Comment (sentilesdal) [reply src/elfpy/pricing_models/yieldspace.py:360]: I went with `mu` and `tau`.  I cleaned a LOT of this up:
`τ`, `t`, `T`, `tau` -> `tau`.
`T`, `t_stretch` -> `t_stretch`
`t` or `t_d` is now only used for normalized time
`μ`, `u` -> `mu`

File: tests/pricing_models/hyperdrive/hd_test_calc_in_given_out.py
Comment (sentilesdal) [reply tests/pricing_models/hyperdrive/hd_test_calc_in_given_out.py]: Added failure tests

File: src/elfpy/pricing_models/hyperdrive.py
Comment (jalextowle) [reply src/elfpy/pricing_models/hyperdrive.py:121]: I'm okay with testing PRs that fix the problems they test; however, I'm fine with this approach in this case.

File: src/elfpy/pricing_models/yieldspace.py
Comment (jalextowle) [inline src/elfpy/pricing_models/yieldspace.py:304]: While we're fixing the use of time in comments, let's do it everywhere. For example this equation and any others you see that still use `t` instead of `\tau`.

File: tests/pricing_models/yieldspace_test_calc_out_given_in.py
Comment (jalextowle) [inline tests/pricing_models/yieldspace_test_calc_out_given_in.py:144]: Nit: `(1 - T)` should have been updated to `(1 - tau)` but it's empty here.

File: tests/pricing_models/yieldspace_test_calc_out_given_in.py
Comment (jalextowle) [inline tests/pricing_models/yieldspace_test_calc_out_given_in.py:210]: I'd recommend that you look through the diffs and find all cases in which `(1 - T)` => `(1 - tau)` and all other changes like this.

File: tests/pricing_models/hyperdrive_test_calc_in_given_out.py
Comment (jalextowle) [inline tests/pricing_models/hyperdrive_test_calc_in_given_out.py:88]: This definitely doesn't need to be fixed in this PR, but I don't like these comments personally. I'm of the opinion that they just add clutter and don't really help to distinguish between the cases. Having said this, I know you're just following the old pattern, so no need to address this here.

**Review by jalextowle — COMMENTED** [review]
Thanks for the hard work on this! This is really starting to come together.

A few more comments here. I am okay with approving once these are addressed, but the tests need to be merged (as one of my earlier comments asked for) in this PR or a follow-up PR. It is really difficult to compare the outputs from different pricing models when they live in different files. If they live together, this issue is trivial _and_ we will significantly reduce code duplication and maintenance.

File: tests/pricing_models/yieldspace_test_calc_out_given_in.py
Comment (sentilesdal) [reply tests/pricing_models/yieldspace_test_calc_out_given_in.py:144]: oops! good catch

File: tests/pricing_models/yieldspace_test_calc_out_given_in.py
Comment (jalextowle) [inline tests/pricing_models/yieldspace_test_calc_out_given_in.py:210]: Nit: This comment isn't right. It should be `(1 - tau)`. I hate to be so nitpicky, but it needs to be done right before we merge it in. If you read through your own diff, you should be able to catch problems like this before review.

Comment (jalextowle) [reply tests/pricing_models/yieldspace_test_calc_out_given_in.py:210]: For things like this, I'd recommend reviewing the changes before you even commit as that way you won't waste your own cycles waiting for review and won't waste the reviewers cycles either.

**Review by jalextowle — APPROVED** [review]
I am approving because it's ready after my last comment is addressed. Having said this, it does need to be addressed, and I'd like you to make sure the comments look correct before merging.

**Pull request merged**
